File: Javascript/Data Structures/Hash Tables/possibleSums.js
function solution(coins, quantity)
{
    var sums = new Set();
    sums.add(0);
    
    for (let i = 0; i < coins.length; i++)
    {
        var next = new Set();

        sums.forEach (function(prev) {
            for (let j = 1; j <= quantity[i]; j++)
            {
                next.add(prev + (coins[i] * j));
            }
        });


        sums.add(next);
    }

    return sums.size - 1;
}

File: Java/Data Structures/src/HashTables/possibleSums.java
package HashTables;

import java.util.HashSet;
import java.util.Set;

public class possibleSums
{
    int solution(int[] coins, int[] quantity)
    {
        Set<Integer> set = new HashSet<>();

        set.add(0);

        for (int i = 0; i < coins.length; i++)
        {
            Set<Integer> next = new HashSet<>();

            for (int prev: set)
            {
                for (int j = 1; j <= quantity[i]; j++)
                {
                    next.add(prev + (j * coins[i]));
                }
            }
            set.addAll(next);
        }

        return set.size() - 1;
    }
}
